yarpcerrors/errors.go

```diff
@@ -25,24 +25,112 @@ import (
 	"fmt"
 )
 
-// IsYARPCError is a convenience function that returns true if the given error
-// is a non-nil YARPC error.
-func IsYARPCError(err error) bool {
-	_, ok := err.(*yarpcError)
+// Newf returns a new Status.
+//
+// The Code should never be CodeOK, if it is, this will return nil.
+func Newf(code Code, format string, args ...interface{}) *Status {
+	if code == CodeOK {
+		return nil
+	}
+	return &Status{
+		code:    code,
+		message: sprintf(format, args...),
+	}
+}
+
+// FromError returns the Status for the error.
+//
+// If the error is a Status, it returns err.
+// If the error is not a Status, it returns a new error with the Code
+// CodeUnknown and the message err.Error()
+func FromError(err error) *Status {
+	if status, ok := err.(*Status); ok {
+		return status
+	}
+	return &Status{
+		code:    CodeUnknown,
+		message: err.Error(),
+	}
+}
+
+// IsStatus returns true if err is a Status.
+func IsStatus(err error) bool {
+	_, ok := err.(*Status)
 	return ok
 }
 
+// Status represents a YARPC error.
+type Status struct {
+	code    Code
+	name    string
+	message string
+}
+
+// WithName returns a new Status with the given name.
+//
+// If s is nil, this will still return nil.
+//
+// This should be used for user-defined errors.
+//
+// The name must only contain lowercase letters from a-z and dashes (-), and
+// cannot start or end in a dash. If the name is something else, an error with
+// code CodeInternal will be returned.
+func (s *Status) WithName(name string) *Status {
```

Comment:
do we want an equivalent "WithCode"

My thought is something like:

`yarpcerrors.FromError(err).WithCode(yarpcerrors.CodeInvalidRequest)`

Reply:
I think we can do that after, let's just have the basics now